File: tests/checkdisk-access.test.ts
/**
 * Exercises the `file access` gate on the disk checks which take a
 * caller-supplied path: check_files, check_single_file and check_disk_write.
 *
 * These never return file contents, but check_files walks whole directory
 * trees and reports every name, size and timestamp, and the checksum keywords
 * turn any readable file into a hash oracle. Where the caller chooses the
 * argument - NRPE with `allow arguments`, or REST - that reach is wide enough
 * to be worth narrowing, and this is the setting that does it.
 *
 * Each case runs a one-shot client query - `nscp client --module CheckDisk
 * --boot --query <cmd> ...` - which re-reads the settings file every time, so
 * the mode can be rewritten between cases with no server to restart. That path
 * also passes `k=v` as single tokens, the same way REST does.
 */
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

import { NscpInstance } from "@fixtures/index";

jest.setTimeout(180_000);

const UNKNOWN = 3;

describe("CheckDisk file access modes", () => {
  let nscp: NscpInstance;
  let scratch: string;
  let allowedDir: string;
  let secretDir: string;
  let allowedLog: string;
  let secretFile: string;

  async function query(command: string, args: string[]): Promise<{ out: string; code: number }> {
    const r = await nscp.run(["client", "--module", "CheckDisk", "--boot", "--query", command, ...args], {
      allowFailure: true,
    });
    return { out: r.all ?? `${r.stdout}\n${r.stderr}`, code: r.exitCode };
  }

  /** Point [/settings/disk] at one mode, always writing both keys. */
  async function setAccess(mode: string, allowed = "-"): Promise<void> {
    await nscp.configure({
      "/settings/disk": { "file access": mode, "allowed files": allowed },
    });
  }

  beforeAll(async () => {
    nscp = new NscpInstance();
    scratch = fs.mkdtempSync(path.join(os.tmpdir(), "checkdisk-access-it-"));
    allowedDir = path.join(scratch, "logs");
    secretDir = path.join(scratch, "secret");
    fs.mkdirSync(path.join(allowedDir, "sub"), { recursive: true });
    fs.mkdirSync(secretDir, { recursive: true });

    allowedLog = path.join(allowedDir, "app.log");
    secretFile = path.join(secretDir, "credentials");
    fs.writeFileSync(allowedLog, "one\ntwo\n");
    fs.writeFileSync(path.join(allowedDir, "sub", "deep.log"), "deep\n");
    fs.writeFileSync(secretFile, "hunter2\n");

    await nscp.configure({ "/modules": { CheckDisk: "enabled" } });
  });

  afterAll(() => {
    fs.rmSync(scratch, { recursive: true, force: true });
  });

  // --- any: the default, and what every earlier release did ------------------

  describe("any (default)", () => {
    beforeAll(() => setAccess("any"));

    it("check_files scans any root the caller names", async () => {
      const { out, code } = await query("check_files", [`path=${secretDir}`, "pattern=*", "detail-syntax=%(filename)", "top-syntax=${list}"]);
      expect(code).not.toBe(UNKNOWN);
      expect(out).toMatch(/credentials/);
    });

    it("check_single_file stats any file the caller names", async () => {
      const { out, code } = await query("check_single_file", [`file=${secretFile}`]);
      expect(code).not.toBe(UNKNOWN);
      expect(out).toMatch(/credentials/);
    });
  });

  // --- allowed: only what matches the list ----------------------------------

  describe("allowed", () => {
    beforeAll(() => setAccess("allowed", allowedDir));

    it("check_files scans a root inside the allowed directory", async () => {
      const { out, code } = await query("check_files", [`path=${allowedDir}`, "pattern=*", "detail-syntax=%(filename)", "top-syntax=${list}"]);
      expect(code).not.toBe(UNKNOWN);
      expect(out).toMatch(/app\.log/);
    });

    it("check_files refuses a root outside the allowed directory", async () => {
      const { out, code } = await query("check_files", [`path=${secretDir}`, "pattern=*", "detail-syntax=%(filename)", "top-syntax=${list}"]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
      expect(out).not.toMatch(/credentials/);
    });

    it("check_files refuses a .. traversal out of the allowed directory", async () => {
      const escape = path.join(allowedDir, "..", "secret");
      const { out, code } = await query("check_files", [`path=${escape}`, "pattern=*", "detail-syntax=%(filename)", "top-syntax=${list}"]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
      expect(out).not.toMatch(/credentials/);
    });

    it("check_files gates the paths= alias too", async () => {
      const { out, code } = await query("check_files", [`paths=${allowedDir},${secretDir}`, "pattern=*"]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
    });

    // Only the scan root is held against the allow list, so `pattern` has to
    // stay a file mask: the Windows scanner concatenates it onto the directory
    // it is walking, and a pattern carrying separators walked straight out of
    // the root which had just passed the gate.
    it("check_files refuses a pattern which climbs out of the scan root", async () => {
      for (const pattern of ["../secret/*", "..\\secret\\*", "../../*", "sub/../../secret/*"]) {
        const { out, code } = await query("check_files", [
          `path=${allowedDir}`,
          `pattern=${pattern}`,
          "detail-syntax=%(filename)",
          "top-syntax=${list}",
        ]);
        expect(code).toBe(UNKNOWN);
        expect(out).toMatch(/Refusing pattern/);
        expect(out).not.toMatch(/credentials/);
      }
    });

    it("check_files refuses a pattern naming a subdirectory while restricted", async () => {
      const { out, code } = await query("check_files", [`path=${allowedDir}`, "pattern=sub/*.log"]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing pattern/);
    });

    it("check_files still accepts an ordinary file mask", async () => {
      const { out, code } = await query("check_files", [
        `path=${allowedDir}`,
        "pattern=*.log",
        "detail-syntax=%(filename)",
        "top-syntax=${list}",
      ]);
      expect(code).not.toBe(UNKNOWN);
      expect(out).toMatch(/app\.log/);
    });

    // A traversal written with the other separator. On Windows it is a
    // separator and must be flattened before the match; on Linux it is an
    // ordinary character and must not become one afterwards. Either way the
    // secret directory must not be enumerated.
    it("check_files does not enumerate through a backslash traversal", async () => {
      const escape = `${allowedDir}/..\\..\\secret`;
      const { out } = await query("check_files", [`path=${escape}`, "pattern=*", "detail-syntax=%(filename)", "top-syntax=${list}"]);
      expect(out).not.toMatch(/credentials/);
    });

    // The checksum keywords are the reason this matters even though no content
    // keyword exists: a hash of an arbitrary file is a content oracle.
    it("check_files cannot hash a file outside the allowed directory", async () => {
      const { out, code } = await query("check_files", [
        `path=${secretDir}`,
        "pattern=*",
        "detail-syntax=%(filename)=%(sha256_checksum)",
        "top-syntax=${list}",
      ]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
      expect(out).not.toMatch(/[0-9a-f]{64}/);
    });

    it("check_single_file refuses a file outside the allowed directory", async () => {
      const { out, code } = await query("check_single_file", [`file=${secretFile}`]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
    });

    it("check_disk_write refuses a path outside the allowed directory", async () => {
      const target = path.join(secretDir, "probe.tmp");
      const { out, code } = await query("check_disk_write", [`file=${target}`, "size=1k"]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
      expect(fs.existsSync(target)).toBe(false);
    });

    it("check_disk_write still works inside the allowed directory", async () => {
      const { code } = await query("check_disk_write", [`file=${path.join(allowedDir, "probe.tmp")}`, "size=1k"]);
      expect(code).not.toBe(UNKNOWN);
    });

    // A symlink is the other way a string comparison would be fooled.
    (process.platform === "win32" ? it.skip : it)("refuses a symlink leading out of the allowed directory", async () => {
      const link = path.join(allowedDir, "escape.log");
      fs.rmSync(link, { force: true });
      fs.symlinkSync(secretFile, link);
      const { out, code } = await query("check_single_file", [`file=${link}`]);
      fs.rmSync(link, { force: true });
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/Refusing file/);
    });

    it("does not disclose the allow list in the refusal", async () => {
      const { out } = await query("check_single_file", [`file=${secretFile}`]);
      expect(out).not.toContain(allowedDir);
    });
  });

  // --- predefined: only names the operator configured ------------------------

  describe("predefined", () => {
    beforeAll(async () => {
      await nscp.configure({ "/settings/disk/files": { logs: allowedDir, app: allowedLog } });
      await setAccess("predefined");
    });

    it("check_files scans a configured name", async () => {
      const { out, code } = await query("check_files", ["path=logs", "pattern=*", "detail-syntax=%(filename)", "top-syntax=${list}"]);
      expect(code).not.toBe(UNKNOWN);
      expect(out).toMatch(/app\.log/);
    });

    it("check_single_file stats a configured name", async () => {
      const { code } = await query("check_single_file", ["file=app"]);
      expect(code).not.toBe(UNKNOWN);
    });

    it("refuses a raw path", async () => {
      const { out, code } = await query("check_single_file", [`file=${allowedLog}`]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/set to predefined/);
    });

    it("resolves a configured name in any mode too", async () => {
      await setAccess("any");
      const { code } = await query("check_single_file", ["file=app"]);
      expect(code).not.toBe(UNKNOWN);
      await setAccess("predefined");
    });
  });

  // --- a typo in the mode must not read as "no restriction" ------------------

  describe("an invalid mode", () => {
    it("fails closed and names the valid values", async () => {
      await setAccess("allwed", allowedDir);
      const { out, code } = await query("check_files", [`path=${allowedDir}`, "pattern=*"]);
      expect(code).toBe(UNKNOWN);
      expect(out).toMatch(/expected any, allowed or predefined/);
      await setAccess("any");
    });
  });
});

File: modules/CheckDisk/check_disk_access_test.cpp
// SPDX-FileCopyrightText: 2004-2026 Michael Medin
// SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-only

// The `file access` gate on the three disk checks which take a caller-supplied
// path: check_files, check_single_file and check_disk_write.
//
// Unlike check_logfile these never return file contents, but check_files walks
// whole directory trees and reports every name, size and timestamp, and the
// checksum keywords turn any readable file into a hash oracle. Where the caller
// chooses the argument - NRPE with `allow arguments`, or REST - that is a wide
// enough reach to be worth narrowing.
//
// The policy itself is covered by include/check/access_policy_test.cpp; what is
// tested here is that each command actually consults it, refuses before doing
// any work, and reports through the normal UNKNOWN path.

#include <check/path_access_policy.hpp>
#include <gtest/gtest.h>

#include <boost/filesystem.hpp>
#include <fstream>
#include <string>

#include "check_disk_write.hpp"
#include "check_files.hpp"
#include "check_single_file.hpp"
#include "test_support.hpp"

namespace {

namespace fs = boost::filesystem;
using check_disk_test_support::join_lines;
using ScratchDir = check_disk_test_support::ScratchDir;

/** A policy restricted to `dir`, in the mode named. */
check::access::path_policy restricted(const std::string &mode, const std::string &allowed) {
  check::access::path_policy policy("file", "files", "/settings/disk");
  policy.set_mode(mode);
  policy.set_allow_list(allowed);
  return policy;
}

PB::Commands::QueryRequestMessage::Request make_request(const std::string &command, const std::vector<std::string> &args) {
  PB::Commands::QueryRequestMessage::Request request;
  request.set_command(command);
  for (const std::string &a : args) request.add_arguments(a);
  return request;
}

class DiskAccessTest : public ::testing::Test {
 protected:
  void SetUp() override {
    allowed_dir_ = (scratch_.path() / "logs").string();
    secret_dir_ = (scratch_.path() / "secret").string();
    fs::create_directories(fs::path(allowed_dir_) / "sub");
    fs::create_directories(secret_dir_);
    write(fs::path(allowed_dir_) / "app.log", "one\ntwo\n");
    write(fs::path(allowed_dir_) / "sub" / "deep.log", "deep\n");
    write(fs::path(secret_dir_) / "credentials", "hunter2\n");
  }
  static void write(const fs::path &p, const std::string &body) {
    std::ofstream f(p.string().c_str());
    f << body;
  }

  ScratchDir scratch_{"nscp-disk-access"};
  std::string allowed_dir_;
  std::string secret_dir_;
};

// ---------------------------------------------------------------------------
// check_files
// ---------------------------------------------------------------------------

TEST_F(DiskAccessTest, CheckFilesAnyModeScansAnyRoot) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"path=" + secret_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, check_disk_test_support::unrestricted());

  EXPECT_NE(response.result(), PB::Common::ResultCode::UNKNOWN) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckFilesAllowsARootInsideTheList) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"path=" + allowed_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_NE(response.result(), PB::Common::ResultCode::UNKNOWN) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckFilesRefusesARootOutsideTheList) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"path=" + secret_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  const std::string out = join_lines(response);
  EXPECT_NE(out.find("Refusing file"), std::string::npos) << out;
  // The refusal must not carry the thing the caller was trying to enumerate.
  EXPECT_EQ(out.find("credentials"), std::string::npos) << out;
}

// The reason a path needs resolving before it is matched: as plain text this
// root reads as "under the allowed directory" and is not.
TEST_F(DiskAccessTest, CheckFilesRefusesATraversalOutOfTheAllowedRoot) {
  PB::Commands::QueryRequestMessage::Request request =
      make_request("check_files", {"path=" + (fs::path(allowed_dir_) / ".." / "secret").string(), "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
}

// `paths=` is a comma separated alias for repeated `path=`; it is split before
// the gate runs, so its entries are held to the same list.
TEST_F(DiskAccessTest, CheckFilesGatesThePathsAlias) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"paths=" + allowed_dir_ + "," + secret_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckFilesRefusesEveryRootWhenOneIsNotAllowed) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"path=" + allowed_dir_, "path=" + secret_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckFilesPredefinedModeResolvesAName) {
  check::access::path_policy policy = restricted("predefined", "");
  policy.add_predefined("logs", allowed_dir_);

  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"path=logs", "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, policy);

  EXPECT_NE(response.result(), PB::Common::ResultCode::UNKNOWN) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckFilesPredefinedModeRefusesARawPath) {
  check::access::path_policy policy = restricted("predefined", "");
  policy.add_predefined("logs", allowed_dir_);

  PB::Commands::QueryRequestMessage::Request request = make_request("check_files", {"path=" + allowed_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response response;

  check_files_command::check(request, &response, policy);

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("set to predefined"), std::string::npos) << join_lines(response);
}

// ---------------------------------------------------------------------------
// check_single_file
// ---------------------------------------------------------------------------

TEST_F(DiskAccessTest, CheckSingleFileAllowsAFileInsideTheList) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_single_file", {"file=" + (fs::path(allowed_dir_) / "app.log").string()});
  PB::Commands::QueryResponseMessage::Response response;

  check_single_file_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_NE(response.result(), PB::Common::ResultCode::UNKNOWN) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckSingleFileRefusesAFileOutsideTheList) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_single_file", {"file=" + (fs::path(secret_dir_) / "credentials").string()});
  PB::Commands::QueryResponseMessage::Response response;

  check_single_file_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  const std::string out = join_lines(response);
  EXPECT_NE(out.find("Refusing file"), std::string::npos) << out;
  // It must refuse rather than leak existence through the "File not found"
  // message the ungated path would produce for an unreadable file.
  EXPECT_EQ(out.find("File not found"), std::string::npos) << out;
}

// The `path=` alias reaches the same variable, so it is gated identically.
TEST_F(DiskAccessTest, CheckSingleFileGatesThePathAlias) {
  PB::Commands::QueryRequestMessage::Request request = make_request("check_single_file", {"path=" + (fs::path(secret_dir_) / "credentials").string()});
  PB::Commands::QueryResponseMessage::Response response;

  check_single_file_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
}

// `ignore-missing` returns OK for a file which is not there; it must not turn a
// refusal into a silent OK, or the gate would be invisible.
TEST_F(DiskAccessTest, CheckSingleFileRefusalBeatsIgnoreMissing) {
  PB::Commands::QueryRequestMessage::Request request =
      make_request("check_single_file", {"file=" + (fs::path(secret_dir_) / "credentials").string(), "ignore-missing=true"});
  PB::Commands::QueryResponseMessage::Response response;

  check_single_file_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
}

#ifndef WIN32
// stat_single_file follows a symbolic link, so resolving the path before the
// match is what stops a link inside an allowed directory reporting on its
// target.
TEST_F(DiskAccessTest, CheckSingleFileRefusesASymlinkLeadingOut) {
  boost::system::error_code ec;
  fs::create_symlink(fs::path(secret_dir_) / "credentials", fs::path(allowed_dir_) / "escape.log", ec);
  if (ec) GTEST_SKIP() << "cannot create symlinks here: " << ec.message();

  PB::Commands::QueryRequestMessage::Request request = make_request("check_single_file", {"file=" + (fs::path(allowed_dir_) / "escape.log").string()});
  PB::Commands::QueryResponseMessage::Response response;

  check_single_file_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
}
#endif

// ---------------------------------------------------------------------------
// check_disk_write
// ---------------------------------------------------------------------------

TEST_F(DiskAccessTest, CheckDiskWriteAllowsAPathInsideTheList) {
  PB::Commands::QueryRequestMessage::Request request =
      make_request("check_disk_write", {"file=" + (fs::path(allowed_dir_) / "probe.tmp").string(), "size=1k"});
  PB::Commands::QueryResponseMessage::Response response;

  check_disk_write_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_NE(response.result(), PB::Common::ResultCode::UNKNOWN) << join_lines(response);
}

TEST_F(DiskAccessTest, CheckDiskWriteRefusesAPathOutsideTheList) {
  const fs::path target = fs::path(secret_dir_) / "probe.tmp";
  PB::Commands::QueryRequestMessage::Request request = make_request("check_disk_write", {"file=" + target.string(), "size=1k"});
  PB::Commands::QueryResponseMessage::Response response;

  check_disk_write_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(response).find("Refusing file"), std::string::npos) << join_lines(response);
  // Refused before anything was created.
  EXPECT_FALSE(fs::exists(target));
}

// The gate runs before the size is parsed, so a refusal is not masked by an
// argument error the caller could use to tell the two apart.
TEST_F(DiskAccessTest, CheckDiskWriteRefusesBeforeValidatingSize) {
  PB::Commands::QueryRequestMessage::Request request =
      make_request("check_disk_write", {"file=" + (fs::path(secret_dir_) / "probe.tmp").string(), "size=not-a-size"});
  PB::Commands::QueryResponseMessage::Response response;

  check_disk_write_command::check(request, &response, restricted("allowed", allowed_dir_));

  EXPECT_EQ(response.result(), PB::Common::ResultCode::UNKNOWN);
  const std::string out = join_lines(response);
  EXPECT_NE(out.find("Refusing file"), std::string::npos) << out;
  EXPECT_EQ(out.find("Invalid size"), std::string::npos) << out;
}

// ---------------------------------------------------------------------------
// a typo in the mode must not read as "no restriction"
// ---------------------------------------------------------------------------

TEST_F(DiskAccessTest, AnInvalidModeFailsClosedOnEveryCommand) {
  const check::access::path_policy policy = restricted("allwed", allowed_dir_);

  PB::Commands::QueryRequestMessage::Request files = make_request("check_files", {"path=" + allowed_dir_, "pattern=*"});
  PB::Commands::QueryResponseMessage::Response files_response;
  check_files_command::check(files, &files_response, policy);
  EXPECT_EQ(files_response.result(), PB::Common::ResultCode::UNKNOWN);
  EXPECT_NE(join_lines(files_response).find("expected any, allowed or predefined"), std::string::npos) << join_lines(files_response);

  PB::Commands::QueryRequestMessage::Request single =
      make_request("check_single_file", {"file=" + (fs::path(allowed_dir_) / "app.log").string()});
  PB::Commands::QueryResponseMessage::Response single_response;
  check_single_file_command::check(single, &single_response, policy);
  EXPECT_EQ(single_response.result(), PB::Common::ResultCode::UNKNOWN);

  PB::Commands::QueryRequestMessage::Request write_req =
      make_request("check_disk_write", {"file=" + (fs::path(allowed_dir_) / "probe.tmp").string(), "size=1k"});
  PB::Commands::QueryResponseMessage::Response write_response;
  check_disk_write_command::check(write_req, &write_response, policy);
  EXPECT_EQ(write_response.result(), PB::Common::ResultCode::UNKNOWN);
}

}  // namespace
